File: src/main/java/io/github/mg138/tsbook/utils/config/gui/armor/ArmorGUIConfig.java
package io.github.mg138.tsbook.utils.config.gui.armor;

import io.github.mg138.tsbook.utils.Translate;
import io.github.mg138.tsbook.utils.config.gui.GUIElementSetting;
import org.bukkit.Material;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.YamlConfiguration;

import java.util.*;

public class ArmorGUIConfig {
    private static ArmorGUIConfig instance = null;

    public static ArmorGUIConfig getInstance() {
        if (instance == null) instance = new ArmorGUIConfig();
        return instance;
    }

    private ArmorGUIConfig() {
    }

    public final Map<Integer, GUIElementSetting> elementSettings = new HashMap<>();

    public void load(YamlConfiguration yaml, Translate translate) {
        elementSettings.clear();

        Set<String> keys = yaml.getKeys(false);
        for (String key : keys) {
            int i = Integer.parseInt(key);

            ConfigurationSection section = yaml.getConfigurationSection(key);
            if (section == null)
                throw new NullPointerException("Something wrong happened when iterating through the config!");

            Material material = Material.valueOf(section.getString("MATERIAL"));

            int count = section.getInt("COUNT");

            String name = translate.translate("NAME", null, section);

            List<String> lore = translate.translateList("LORE", null, section);
            lore = lore == null ? Collections.emptyList() : lore;

            int model = section.getInt("MODEL");

            GUIElementSetting elementSetting;
            if (section.contains("EQUIPMENT")) {
                ConfigurationSection equipment = section.getConfigurationSection("EQUIPMENT");
                if (equipment == null)
                    throw new NullPointerException("Something wrong happened when iterating through the config!");

                elementSetting = new ArmorGUIElementSetting(i, material, count, name, lore, model, new ArmorSetting(equipment));
            } else {
                elementSetting = new GUIElementSetting(i, material, count, name, lore, model);
            }

            elementSettings.put(i, elementSetting);
        }
    }
}

File: src/main/java/io/github/mg138/tsbook/setting/gui/armor/ArmorGUIConfig.kt
package io.github.mg138.tsbook.setting.gui.armor

import io.github.mg138.tsbook.players.ArcticGlobalDataService
import io.github.mg138.tsbook.players.data.PlayerData
import io.github.mg138.tsbook.setting.gui.armor.element.ArmorElementSetting
import io.github.mg138.tsbook.setting.gui.armor.element.ArmorSetting
import io.github.mg138.tsbook.setting.gui.element.GUIElementSetting
import io.github.mg138.tsbook.utils.Translate
import org.bukkit.Material
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.util.*

object ArmorGUIConfig {
    private const val npeMessage = "Something wrong happened when iterating through the config!"
    private val elementSettings: MutableMap<Int, GUIElementSetting> = HashMap()

    operator fun get(key: Int) = elementSettings[key]

    fun getByType(type: String, player: Player): Int {
        val data = ArcticGlobalDataService.inst.getData<PlayerData>(player) ?: return -1

        elementSettings.forEach { (i, setting) ->
            if (!data.equipment.containsKey(i)) {
                if (setting is ArmorElementSetting) {
                    if (setting.setting.type == type) return i
                }
            }
        }

        return -1
    }

    fun load(yaml: YamlConfiguration, translate: Translate) {
        elementSettings.clear()
        for (key in yaml.getKeys(false)) {
            val i = key.toInt()
            val section = yaml.getConfigurationSection(key) ?: throw NullPointerException(npeMessage)

            val material = Material.valueOf(section.getString("MATERIAL")!!)
            val count = section.getInt("COUNT")
            val name = translate.translate("NAME", null, section)
            val lore = translate.translateList("LORE", null, section)
            val model = section.getInt("MODEL")

            elementSettings[i] = when {
                section.contains("EQUIPMENT") -> {
                    val equipment = section.getConfigurationSection("EQUIPMENT") ?: throw NullPointerException(npeMessage)
                    ArmorElementSetting(i, material, count, name, lore, model, ArmorSetting(equipment))
                }
                else -> GUIElementSetting(i, material, count, name, lore, model)
            }
        }
    }

    fun unload() {
        elementSettings.clear()
    }
}